File: 2014/J143.cpp
#include <cstdio>
int a[50000];
int i;
int serial(int x, int y) {
  return x * 50000 + y;
}
bool search(int x, int y) {
  int z = serial(x, y);
  int l = 0, r = i;
  while (r >= l) {
    int mid = (l + r) / 2;
    if (a[mid] == z) {
      return true;
    } else if (a[mid] > z) {
      r = mid - 1;
    } else {
      l = mid + 1;
    }
  }
  return false;
}
int main() {
  int w, h, n, x, y;
  int count = 0;
  scanf("%d %d %d", &w, &h, &n);
  for (i = 0; i < n; i++) {
    scanf("%d %d", &x, &y);
    a[i] = serial(x, y);
    // upright
    if (search(x - 4, y) && search(x - 3, y) && search(x - 2, y + 3) && search(x - 2, y + 2) && search(x - 1, y))
      count++;
    // T
    if (search(x - 4, y) && search(x - 3, y) && search(x - 2, y - 3) && search(x - 2, y - 2) && search(x - 1, y))
      count++;
    // -|
    if (search(x - 3, y - 2) && search(x - 2, y - 2) && search(x, y - 4) && search(x, y - 3) && search(x, y - 1))
      count++;
    // |-
    if (search(x - 3, y - 2) && search(x - 3, y - 1) && search(x - 3, y + 1) && search(x - 3, y + 2) && search(x - 1, y))
      count++;
  }
  printf("%d\n", count);
  return 0;
}

File: 2014/J143.pas
var
  a: array[1..50000] of longint;
  i, w, h, n, x, y, count: longint;
function serial(x, y: longint): longint;
begin
  serial := x * 50000 + y;
end;
function search(x, y: longint): boolean;
var
  l, r, mid, z: longint;
begin
  search := false;
  z := serial(x, y);
  l := 1;
  r := i;
  while (r >= l) do
  begin
    mid := (l + r) div 2;
    if (a[mid] = z) then
    begin
      search := true;
      exit
    end
    else if (a[mid] > z) then
      r := mid - 1
    else
      l := mid + 1
  end
end;
begin
  count := 0;
  read(w, h, n);
  for i := 1 to n do
  begin
    read(x, y);
    a[i] := serial(x, y);
    if (search(x - 4, y) and search(x - 3, y) and search(x - 2, y + 3) and search(x - 2, y + 2) and search(x - 1, y)) then
      inc(count);
    if (search(x - 4, y) and search(x - 3, y) and search(x - 2, y - 3) and search(x - 2, y - 2) and search(x - 1, y)) then
      inc(count);
    if (search(x - 3, y - 2) and search(x - 2, y - 2) and search(x, y - 4) and search(x, y - 3) and search(x, y - 1)) then
      inc(count);
    if (search(x - 3, y - 2) and search(x - 3, y - 1) and search(x - 3, y + 1) and search(x - 3, y + 2) and search(x - 1, y)) then
      inc(count);
  end;
  writeln(count)
end.
